app/mempool/lane.go

```diff
@@ -0,0 +1,272 @@
+	signerExtractor sdkmempool.SignerExtractionAdapter,
+	name string,
+	matchFn func(sdk.Context, sdk.Tx) bool,
+	maxTransactionSpace math.LegacyDec,
+	maxLaneSpace math.LegacyDec,
+	laneMempool sdkmempool.Mempool,
+) *Lane {
+	return &Lane{
+		logger:              logger,
+		txEncoder:           txEncoder,
+		signerExtractor:     signerExtractor,
+		name:                name,
+		matchFn:             matchFn,
+		maxTransactionSpace: maxTransactionSpace,
+		maxLaneSpace:        maxLaneSpace,
+		laneMempool:         laneMempool,
+
+		// Initialize the txIndex.
+		txIndex: make(map[string]struct{}),
+	}
+}
+
+// Insert inserts a transaction into the lane's mempool.
+func (l *Lane) Insert(ctx context.Context, tx sdk.Tx) error {
+	txInfo, err := l.GetTxInfo(tx)
+	if err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if err = l.laneMempool.Insert(ctx, tx); err != nil {
+		return err
+	}
+
+	l.txIndex[txInfo.Hash] = struct{}{}
+	return nil
+}
+
+// CountTx returns the total number of transactions in the lane's mempool.
+func (l *Lane) CountTx() int {
+	return l.laneMempool.CountTx()
+}
+
+// Remove removes a transaction from the lane's mempool.
+func (l *Lane) Remove(tx sdk.Tx) error {
+	txInfo, err := l.GetTxInfo(tx)
+	if err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if err = l.laneMempool.Remove(tx); err != nil {
+		return err
+	}
+
+	delete(l.txIndex, txInfo.Hash)
+	return nil
+}
+
+// Contains returns true if the lane's mempool contains the transaction.
+func (l *Lane) Contains(tx sdk.Tx) bool {
+	txInfo, err := l.GetTxInfo(tx)
+	if err != nil {
+		return false
+	}
+
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+
+	_, exists := l.txIndex[txInfo.Hash]
+	return exists
+}
+
+// Match returns true if the transaction belongs to the lane.
+func (l *Lane) Match(ctx sdk.Context, tx sdk.Tx) bool {
+	return l.matchFn(ctx, tx)
+}
+
+// FillProposal fills the proposal with transactions from the lane mempool with its own limit.
+// It returns the total size and gas of the transactions added to the proposal.
+// It also returns an iterator to the next transaction in the mempool and a list
+// of transactions that were removed from the lane mempool.
+func (l *Lane) FillProposal(
+	ctx sdk.Context,
+	proposal *Proposal,
+) (blockUsed BlockSpace, iterator sdkmempool.Iterator, txsToRemove []sdk.Tx) {
+	// Get the transaction and lane limit for the lane.
+	transactionLimit := proposal.GetLimit(l.maxTransactionSpace)
+	laneLimit := proposal.GetLimit(l.maxLaneSpace)
+
+	// Select all transactions in the mempool that are valid and not already in the
+	// partial proposal.
+	for iterator = l.laneMempool.Select(ctx, nil); iterator != nil; iterator = iterator.Next() {
+		// If the total size used or total gas used exceeds the limit, we break and do not attempt to include more txs.
+		// We can tolerate a few bytes/gas over the limit, since we limit the size of each transaction.
+		if laneLimit.IsReachedBy(blockUsed) {
+			break
+		}
```

Comment:
Why not check this at the end of loop?

Reply:
I think check first is better na, it means that if the lane is at the limit we don't put any more tx.
we can imagine the extreme case of lane limit of 0, check first won't put any tx in the lane but check last will.